src/frontend/src/scheduler/hummock_snapshot_manager.rs

```diff
@@ -144,49 +155,37 @@ struct HummockSnapshotManagerCore {
     epoch_to_query_ids: BTreeMap<u64, HashSet<QueryId>>,
     meta_client: Arc<dyn FrontendMetaClient>,
     last_unpin_snapshot: Arc<AtomicU64>,
+    max_committed_epoch: Arc<AtomicU64>,
 }
 
 impl HummockSnapshotManagerCore {
-    fn new(meta_client: Arc<dyn FrontendMetaClient>) -> Self {
+    fn new(meta_client: Arc<dyn FrontendMetaClient>, max_committed_epoch: Arc<AtomicU64>) -> Self {
         Self {
             // Initialize by setting `is_outdated` to `true`.
             meta_client,
             epoch_to_query_ids: BTreeMap::default(),
             last_unpin_snapshot: Arc::new(AtomicU64::new(0)),
+            max_committed_epoch,
         }
     }
 
-    async fn get_epoch_for_query(
+    fn get_epoch_for_query(
```

Comment:
So now for each batch query, we will never trigger a get_epoch RPC from FE but only get the `max_committed_epoch` from memory? Previously @Little-Wallace and I thinks that we can use 2 consistency mode to choose, so that we can trade off between perf and correctness. 

But indeed, now the flush response contains epoch, so we can partially solve the correctness problem.

Another question, how we handle epoch unpin now?

Reply:
> So now for each batch query, we will never trigger a get_epoch RPC from FE but only get the max_committed_epoch from memory?

Yes

> partially solve the correctness problem

Why partially?

> Another question, how we handle epoch unpin now?

The unpin logic remains the same. But since you mention it, I think we can check release epoch and locally cached max epoch to make sure the pushing is working.

Reply:
If we want the async version for stronger consistency in the future, then we should keep the original implementation.

Reply:
> Why partially?

User may forget to use flush. Also I'm not sure whether we export this feature to user to be honest..

Reply:
But this PR LGTM

Reply:
> > Why partially?
> 
> User may forget to use flush. Also I'm not sure whether we export this feature to user to be honest..

If user forgets to use flush, we cannot guarantee read-after-write prior to this PR either because nothing prevents us from processing a read query when the checkpoint of the previously inserted data is still in-flight.

Reply:
> > > Why partially?
> > 
> > 
> > User may forget to use flush. Also I'm not sure whether we export this feature to user to be honest..
> 
> If user forgets to use flush, we cannot guarantee read-after-write prior to this PR either because nothing prevents us from processing a read query when the checkpoint of the previously inserted data is still in-flight.

In real production, I guess the read and write will be 2 path, so there is no-room for user to type "flush". But Indeed it can "partially solve" because devs can ensure read-after-write.

So if we want strong-consistency in production we might still need rpc path. But there is another problem: In such a streaming environment, will user care so much on consistency, or what is the real consistency? That's why I think this PR is also ok for now stage. We can add support when we really need this.

Reply:
So I'm open to delete some code relates to get_epoch/pin_epoch RPC.... I do not see any cases we very need it..

Reply:
Yep. I'm keeping the original rpc method.

Reply:
> Yep. I'm keeping the original rpc method.

Maybe mark get_epoch_for_query_from_rpc as clippy expect unused is better 🐶  and use get_epoch_for_query for now